### DISCUSSION TITLE
Bleach is deprecated

### DISCUSSION THREAD

**Discussion by oscarmcm** [django-wiki/django-wiki#1259] [Bug] [answered]
Recently Mozilla announced that the project `Bleach` is deprecated in the mozilla/bleach#698. As we know `Django-Wiki` uses it internally and we recently had some issues with a breaking change introduced in the newest version via #1252 and #1257 that were solved via #1253

Maybe it's a good moment to review a better alternative or check how much we use it and maybe just modularize internally.

**Comment by oscarmcm** [comment]
Seems that Weblate is switching to `nh3` as per WeblateOrg/weblate#8664 - Maybe it's a good option or alternative.

**Reply by benjaoming** [reply]
I dunno. [When we went for Bleach](https://github.com/django-wiki/django-wiki/issues/571), the intention was to find something that would be stable for many many years. The Python binding for nh3 may or may not be a stable choice.

**Reply by benjaoming** [reply]
nh3 is now so widely recognized that Adam Johnson has written a blog post about it :)

https://adamj.eu/tech/2023/12/13/django-sanitize-incoming-html-nh3/

**Comment by oscarmcm** [comment, marked as answer]
Please check the following comments in the merge request #1254 

- https://github.com/django-wiki/django-wiki/pull/1254#issuecomment-1406693290
- https://github.com/django-wiki/django-wiki/pull/1254#issuecomment-1406759270

As stated above, we'll update Django-Wiki Bleach dependency to the latest supported version and wait until an alternative comes in.

This discussion will be closed, please open a new one if a solid alternative exists or if the Bleach version supported is causing some issues.

**Comment by benjaoming** [comment]
While this conversation is open, it would also be interesting to discuss the implications of introducing markdown-it-py: https://markdown-it-py.readthedocs.io/en/latest/

It's just gone through a security review by Ubuntu's engineers. And been packaged for Debian. It's a very interesting project and potentially richer to extend from than python-markdown in the longer run. Would have to look at the details, though.

**Reply by oscarmcm** [reply]
Seems to be very promising, maybe instead of choosing one or another we can create a good wrap so then people can use `mistune`, `markdown-it-py`, or `py-markdown` as just plug n' play.

**Reply by benjaoming** [reply]
It's a bit complex because of the plugin layer. I'm thinking that the way plugins behave quickly becomes dependent on what Markdown backend the wiki has chosen.

**Reply by oscarmcm** [reply]
So the more complex part is to have the plugins not be really attached to a specific lib, what about using regex on top of code blocks?

something like this:

```txt
    ```wiki
    [image_1 left]
        some caption
    ```
```

**Reply by benjaoming** [reply]
That's a good idea! We could offer plugins a way to extend Markdown that goes through django-wiki's own abstraction. That's also nice because it can introduce a uniform syntax and not leave it up to each plugin to do whatever they want.

I guess step no. 1 would be to create that abstraction for python-markdown :)

**Reply by benjaoming** [reply]
> what about using regex on top of code blocks?

I guess we need to figure out a pattern for supporting ways of extending Markdown that are offered by each Markdown library and create an abstraction on top of that.

But there are some definite problems here. The most notable one is if we want to be able to operate on the syntax tree... this is too detailed to create an abstraction layer for, in that case we might as well just vendor an entire Markdown library :)

An example of this problem is when an abstraction should choose if it's a pre-processor (replaces markdown with new markdown) or a post-processor (replaces markdown with html). I only know these concepts from Python-Markdown, so I don't know if you can plugin to markdown-it-py the same way.

The issue is: Should a markdown tag be able to replace itself with new markdown tags that are then parsed. For instance, do you want to be able to have a markdown table in an image caption? Do you want to have another image inside an image caption? Etc...


```python
class AbstractMarkdownTag:
    """Abstract base class for all markdown tags [tag_name arg1 kwarg2="foobar"]"""
    class __init__(self, tag_name, markdown_args, markdown_kwargs, *args, **kwargs):
        pass

class PyMarkdownMarkdownTag(AbstractMarkdownTag):
    # ...

# Registry for functions that we want to call
tag_callbacks = {}

def register_markdown_tag(callback_renderer, tag_name, markdown_args, markdown_kwargs):

    if is_python_markdown():
        tag=PyMarkdownMarkdownTag(tag_name, markdown_args, markdown_kwargs)

    elif is_markdown_it_py():
        tag=MarkdownItPyMarkdownTag(...)

    # Register the callback function so that when we are rendering the tag, we can look
    tag_callbacks[tag_name] = callback_renderer


```

Then in the plugin:

```
class SomeImagePlugin():

    def image_tag_callback(self, markdown_instance, article_revision, tag_name, tag_args, tag_kwargs):
        """Renders the Markdown code that should replace the tag. But what code? More markdown to be parsed? Or HTML? This is a difficult problem"""
        pass


```

So like, after thinking about it, I don't think it's possible to support several markdown libraries with an abstraction layer... I think that might be Mission Impossible :)
